Update refactor_business_logic_more_functional.md

MichaelViveros opened the pull request.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+I recently took an excellent Elixir course called ["Elixir for Programmers"](https://codestool.coding-gnome.com/courses/elixir-for-programmers) by Dave Thomas (who my young, millenial self learned was one of the authors of the timeless book entitled The Pragmatic Programmer). One of the principles emphasized by the course is the importance of keeping your application/business logic separate from  ...
+
+The code base I set my sights on was for the new version of PagerDuty's webhook-processing service. The code for our existing webhook-processing service is relatively old (5+ years) and written in Scala. There has been a growing movement within PagerDuty to [prefer Elixir](https://www.pagerduty.com/blog/elixir-at-pagerduty/) over other languages like Scala so my team has started re-writing that s ...
```
The correct wording here is "PagerDuty standardized on Elixir last year". May be nitpicking but when we present ourselves we should be as precise as possible because otherwise we will have a lot of raised eyebrows with candidates.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+The code base I set my sights on was for the new version of PagerDuty's webhook-processing service. The code for our existing webhook-processing service is relatively old (5+ years) and written in Scala. There has been a growing movement within PagerDuty to [prefer Elixir](https://www.pagerduty.com/blog/elixir-at-pagerduty/) over other languages like Scala so my team has started re-writing that s ...
+
+The code for our shiny new Elixir webhook-processing service was originally written by people who were/are quite new to Elixir (myself included) and it followed more of a traditional approach of having your business logic spread across lots of different parts of your application. This made it hard to get a big picture view of our service's business logic without travelling to the deep, dark depth ...
```
"traditional" where? I think it's a very bad approach everywhere...

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+The code for our shiny new Elixir webhook-processing service was originally written by people who were/are quite new to Elixir (myself included) and it followed more of a traditional approach of having your business logic spread across lots of different parts of your application. This made it hard to get a big picture view of our service's business logic without travelling to the deep, dark depth ...
+
+This was not very functional, even though it was written in a functional language. So I refactored it to be more functional.
```
I don't think that bad coding practices have anything to do with functional versus non-functional...

cdegroot replied:
The two preceding paragraphs probably are trying to tell that the team was new to the language and therefore the result wasn't really nice and could use some cleanup. I'd leave out the whole "functional" bit.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+Webhooks get generated by an upstream service whenever an incident is triggered/acknowledged/resolved. This upstream service sends webhooks to our webhook-processing service which sends out these webhooks to their endpoints.
+
+In addition to this simple business logic, we have more business logic to deal with things like expired webhooks that are older than 5s which we should not send out since for our customers, old data is just as bad, maybe even worse, than no data. Imagine getting paged about a mysterious API error and then, after 15 mins of investigation and pinpointing the issue to one of your MySQL hosts, you ge ...
```
I'd use shorter sentences here, they don't really flow well.

cdegroot replied:
(also, don't overuse the term "business logic", find some alternative wordings)

bzmw replied:
The 5s part isn't true, as we discovered this morning ;)

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+In addition to this simple business logic, we have more business logic to deal with things like expired webhooks that are older than 5s which we should not send out since for our customers, old data is just as bad, maybe even worse, than no data. Imagine getting paged about a mysterious API error and then, after 15 mins of investigation and pinpointing the issue to one of your MySQL hosts, you ge ...
+
+We also have some business logic to save the responses returned from the webhook endpoints into a database that we can use to help debug issues with customer endpoints (which sometimes honestly feel like they send us back more 400s/500s than 200s and then their engineers, who forgot to whitelist our webhook-processing hosts' IPs, come yelling at us ...).
```
Please don't picture customer staff yelling at us in external blog posts :)

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+Enough talk, here's some code.
+
+(Note that the libraries being used in the code aren't actual elixir libraries, I preferred simplicity/readability over accuracy/runnability. Similarly, the overall webhook-processing business logic is vastly different from our actual business logic.)
```
I don't understand either remark here. I guess that the first one is trying to tell us that this is not our actual code, and the second one probably the same? Maybe replace it with "this is a simplified form of our actual code to help illustrate the point of this article better"?

rachfrieee replied:
I'd remove the parenthesis and rewrite the intro to the sentence slightly so that it stands on its own.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+Here's a diagram of the business logic and how it's spread out over different modules:
+![Old Business Logic](OldBusinessLogic.png)
+
+In order to peice together that diagram above, you have to start in `WebhookReceiver`, go to `WebhookParser`, go back to `WebhookReceiver`, go to `WebhookSender` and finally finish off in `WebhookDB`.
```
s/peice/piece/

bzmw replied:
This sentence highlights your overuse of the word `Webhook` in your modules naming. It would be much easier to refer to your modules if you just dropped `Webhook` from all of them, and it would be easier to follow along as the reader.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+In order to peice together that diagram above, you have to start in `WebhookReceiver`, go to `WebhookParser`, go back to `WebhookReceiver`, go to `WebhookSender` and finally finish off in `WebhookDB`.
+
+Along the way, you encounter some business logic in modules that seem totally unrelated to the business logic they are executing. For example, the logic about not processing a webhook if it expired is contained in the `WebhookParser` module whose sole responsibility should be decoding the webhook from JSON. Similarly, the logic about storing different types of webhook endpoint responses into the  ...
```
I'd write "...totally unrelated to the role they are supposed to play" to avoid "logic"/"business logic". But I already remarked on the overuse of the term ;-)

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+Now, overall, this business logic wasn't too hard to figure out for the simplified code above but for an application with any reasonable level of complexity, this can be quite daunting.
+
+This also does not take advantage of the fact that webhook processing can be thought of as a pipeline which can be expressed beautifully in functional programming languages like Elixir where you take in some initial state and keep modifying it through different stages to produce an output.
```
It's not super clear what "This" refers to. I'd say "The code above" or something to make it clearer.

cdegroot commented on refactor_business_logic_more_functional.md:
````diff
@@ -1 +1,287 @@
+
+```elixir
+defmodule WebhookProcessor do
+  defstruct(
````
Usually, if I see that I expect a `%WebhookProcessor{...}` to appear somewhere but it doesn't. Also, the default name is probably not what you want for that struct, and in `process_webhook` you seem to convolute the webhook's state with the actual webhook? At the very least, define the struct in a submodule with the correct name (`State` or `WebhookState` or even - probably most correct - `WebhookWithState`?) and then type your arguments a bit more in the code by referencing the struct; without that defining it doesn't really make sense.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+        # now we call WebhookProcessor.process instead of WebhookSender.send
+
+        %WebhookProcessor{
```
Ah, here the struct is used and here it hopefully becomes clear that `WebhookProcessor` is not a really good name for the type :-)

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+So this is how I refactored the business logic behind our webhook-processing pipeline to be more functional.
+
+Recall some of the problems with having your business logic spread out across different modules:
+* Takes a long time to move through all these different modules and understand wtf is going on
```
Make these bullet points full sentences, I'd say. Easier to read.

bzmw replied:
I think this is the first time you used an acronym such as 'wtf', either use more of them or remove this one usage.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+* Can clearly see what state your business logic cares about and how it modifies that state in the pipeline
+* I didn't mention this but it also makes it easier to unit test your business logic
+
+Now I would be remiss not to mention that there are some benefits to embedding your business logic within different modules of your app. This can lead to less code and faster development and is one of the fundamental ideas behind frameworks like Ruby On Rails that prefer "convention over configuration". I personally disagree with this but I wouldn't go as far to say it is "flat-out wrong". Hopefu ...
```
This is actually flat-out wrong. I'd leave it out, or call it out what it is: a horribly bad, flat out wrong idea.

bzmw commented on refactor_business_logic_more_functional.md:
````diff
@@ -1 +1,287 @@
+```
+
+Here's a diagram of the business logic and how it's spread out over different modules:
+![Old Business Logic](OldBusinessLogic.png)
````
Reading this it seemed very disjointed. Some comments in the code to explain them would be helpful.

To me it felt like going from a conversation with you to reading some random persons' code.

bzmw commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+## The New Code
+
+So to centralize our business logic into one module and to express it as a pipeline, I moved all the bits of business logic spread throughout the code into a new module called `WebhookProcessor`.
```
Overusing 'business logic'

bzmw commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+All of our business logic is now contained within `WebhookProcessor` and instead of jumping around modules to see it you only have to scroll up and down in one module.
+
+We also now have a nice pipeline defined in `process_webhook` where we first check if the webhook expired. If it expired, we return an error and stop the pipeline so the webhook does not get sent out. If it isn't expired, then we move onto the next stage where we send it to its endpoint. Then we handle the response returned by the endpoint and then save it to the database.
```
Starting in this paragraph you changed tenses to "We" instead of "I"

I'd suggest going with "I" throughout because your writing style speaks as you do.

bzmw commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+And here's a diagram of the refactored business logic which is all contained in the `WebhookProcessor` module which calls out to the other modules:
+![New Business Logic](NewBusinessLogic.png)
+
+Once again, I wanted to lightly remind you that the code above is a simplification of our actual webhook processing pipeline. If I see any support requests (or even better, *tweets*) saying something like "Do you actually not handle our endpoints returning status codes besides 200 or 404???" or "Does PagerDuty seriously not even bother to retry sending failed webhooks???", I will simply smh. Then ...
```
This is cute, however a bit long winded.

rachfrieee commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+(0 mins if you [disagree](https://twitter.com/dhh/status/218740829806792704) ;))
+
+I recently took an excellent Elixir course called ["Elixir for Programmers"](https://codestool.coding-gnome.com/courses/elixir-for-programmers) by Dave Thomas (who my young, millenial self learned was one of the authors of the timeless book entitled The Pragmatic Programmer). One of the principles emphasized by the course is the importance of keeping your application/business logic separate from  ...
```
I'm betting we could add a link to the book on Amazon

writerevie replied:
Formatting nit (and typo fix): (who my young, millennial self learned was one of the authors of the timeless book, "The Pragmatic Programmer").

rachfrieee commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
 
+# Functional Programming and Business Logic, Via Elixir
+
+10 min read
```
For the PD blog, I'd pull the first 3 lines about the read length -- just b/c we don't stylistically we don't include that on the PD blog

rachfrieee commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+
+Along the way, you encounter some business logic in modules that seem totally unrelated to the business logic they are executing. For example, the logic about not processing a webhook if it expired is contained in the `WebhookParser` module whose sole responsibility should be decoding the webhook from JSON. Similarly, the logic about storing different types of webhook endpoint responses into the  ...
+
+Now, overall, this business logic wasn't too hard to figure out for the simplified code above but for an application with any reasonable level of complexity, this can be quite daunting.
```
I'd remove "Now," and just start the sentence with "Overall"

rachfrieee commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,287 @@
+* Why PagerDuty [<3](https://www.pagerduty.com/blog/elixir-at-pagerduty/) Elixir
+* PagerDuty Webhooks - [Overview](https://v2.developer.pagerduty.com/docs/webhooks-v2-overview), [Setup Guide](https://support.pagerduty.com/docs/webhooks)
+* Hack Days @ PagerDuty - [Link 1](https://www.pagerduty.com/blog/pagerduty-april-hack-day/), [Link 2](https://www.pagerduty.com/blog/create-awesome-engineering-culture/) (Point #2)
+* [Careers](https://www.pagerduty.com/careers/) at PagerDuty where you're empowered to do things like refactoring your code to be more functional
```
I'm wondering if we couldn't also include a plug for the community forums as a place to continue the discussion, thoughts on this?

rachfrieee replied:
Oh perfect! We can just update it once the blog goes live. Good thinking!

MichaelViveros commented:
Changed the intro and added a section about `Did This Refactor Actually Work?`. It made the post a bit longer but this will hopefully make it more interesting to engineers passing by in the wild.

MichaelViveros commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+One of the principles emphasized by the course is the importance of keeping your business logic separate from everything else, especially your framework. This really resonated with me, so much so that for my next [Hack Day](https://www.pagerduty.com/blog/pagerduty-april-hack-day/) I refactored the Elixir code base mentioned above to adhere to this principle. What resulted, was one *divine* module ...
+
+![New Business Logic](NewBusinessLogic.png)
```
Is it ok to preview the end result or is it better to just leave it until the end?

morriswchris replied:
I've seen a few ways of doing this:

* reference that the full source for the post can be found <insert github repo here>
* Link to the final product at the bottom
* show the preview here

 I typically like the approach of "you can view the code here" as opposed to a large blob of code I need to scroll past.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+Ok, that might be an overexaggeration but my team (the software one, not the hockey one) found ourselves in a similar spot a couple weeks ago. We finished our latest project and were starting a new one which involved rewriting an old (5+ years) Scala service into Elixir. This service backed PagerDuty's [webhook-processing system](https://v2.developer.pagerduty.com/docs/webhooks-v2-overview). 
+
+The decision to undertake this project as our next one was made for 2 main reasons:
+* To help the company [standardize on Elixir](https://www.pagerduty.com/blog/elixir-at-pagerduty/) due to an org-wide decision made last year
```
Nit: "last year" can, I think, refer to the past part of the current year or the previous year. I'd say "earlier this year" to be unambiguous.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+Forgive me for the oozing sarcasm, it's my way of coping.
+
+Now this rewrite project had actually started a year ago but was put on pause. What resulted was a relatively small Elixir code base that my team could use as a starting point for our glorious rewrite. Now this code base wasn't in nearly as bad a spot as our 5+ year-old code but it was also originally written by people who were/are quite new to the language, Elixir (I'm including myself as one of ...
```
Style: starting two sentences in a row with "Now". I think you can drop both instances without hurting the sentences (less is more in writing).

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+Now this rewrite project had actually started a year ago but was put on pause. What resulted was a relatively small Elixir code base that my team could use as a starting point for our glorious rewrite. Now this code base wasn't in nearly as bad a spot as our 5+ year-old code but it was also originally written by people who were/are quite new to the language, Elixir (I'm including myself as one of ...
+
+It followed the *convenient* approach of spreading your business logic across lots of different parts of your application. This made it hard for new team members to get a big picture view of the service's application logic without travelling to the deep, dark depths of Mordor to figure out that the code that expires stale webhooks is embedded in the `Parser` module whose sole responsiblity should ...
```
I'd use "the application" here.

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+Along the way, you encounter some fundamental logic about how webhooks get processed which is hidden in modules that seemingly have nothing to do with this fundamental logic. As an example, the logic about not processing a webhook if it expired is contained in the `Parser` module whose sole responsibility should be decoding the webhook from JSON. Similarly, the logic about storing different types ...
+
+Overall, this webhook-processing logic wasn't too hard to figure out for the simplified code above but for an application with any reasonable level of complexity, this can be quite daunting.
```
I'd make this more concrete to pertain to actual WDS. For example, "Overall, this webhook processing logic wouldn't be too hard to figure out for the simplified code above but for our application, which had a lot of additional complexity, it was quite daunting". Regardless, I think you can drop the dash in "webhook processing" ;-)

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+Overall, this webhook-processing logic wasn't too hard to figure out for the simplified code above but for an application with any reasonable level of complexity, this can be quite daunting.
+
+The code above also does not take advantage of the fact that webhook processing can be thought of as a pipeline which can be expressed beautifully in functional programming languages where you take in some initial state and keep modifying it through different stages to produce an output. Elixir implements a simple version of these pipelines via it's [pipe operator (`|>`)](https://elixirschool.com ...
```
Ah! Found one! Just realised I didn't spot a single instance of the "it's" vs "its" mix-up :)

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+So with the new code, all of the webhook-processing logic is now contained within `Processor` and instead of jumping around modules to see it you only have to look in one module.
+
+There is also a simple pipeline defined in `process` where it first checks if the webhook expired. If the webhook is expired, an error is returned and the pipeline is stopped so the webhook does not get sent out. If it isn't expired, then the webhook is sent to its endpoint in the next stage. Then the different responses returned by the endpoint are handled and saved to the database.
```
"whether the webhook expired", maybe?

cdegroot commented on refactor_business_logic_more_functional.md:
````diff
@@ -1 +1,412 @@
+|> handle_response
+|> save_response_to_db
+```
+Those pipelines are great when you have a simple happy path that given some initial state, modifies it through different stages and produces an output. But they're not well-suited for pipelines like webhook-processing where any stage could throw an error and cause you to stop the pipeline immediately. In the functional programming world, this is known as [Railway Oriented Programming](https://fsh ...
````
So here it gets tricky. In ROP, I think, you'd still have a straight pipeline but every function returns a monadic value and accepts one. For example:

    def check_webhook_expired, do: {:ok, :some_answer}

to start the pipeline, and then

    def send_webhook({:error, _} = error), do: error
    def send_webhook({:ok, state}), do: <some processing>


etcetera which will allow you to write the pipeline with the pipe operator. Error monads are simply passed down and only handled in the final stage. Whether `with` or a ROP style pipeline are better, is a matter of taste I think. Both forms are largely equivalent but `with` is a bit more explicit.

MichaelViveros replied:
Gotcha, thx, I'll stick with `with` since it's cleaner

cdegroot commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+So have I inspired you to drop everything and get refactoring? Or do you vehemently disagree with keeping your business logic separate from everything else? All feedback and discussion is welcome in our PagerDuty Community [forum](https://community.pagerduty.com/c/blog)!
+
+If you'd be interested in working for a company that empowers it's engineers to do things like refactoring unfamiliar code to improve their team's confidence with it, check out our [current job openings](https://www.pagerduty.com/careers/).
```
Ah, found the second instance of "it's"!

Tip: don't use "it's" in writing. Use "it is". Then you can basically find all the occurrences where you're muscle memory made you type "it's" and fix them. 

[and yes, I made that error in the previous sentence on purpose. As a pun, y'know]

morriswchris replied:
I hope that careers link isn't your referral one 😉

morriswchris commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+![New Business Logic](NewBusinessLogic.png)
+
+So with the new code, all of the webhook-processing logic is now contained within `Processor` and instead of jumping around modules to see it you only have to look in one module.
```
style: I think the `and` is redundant

> So with the new code, all of the webhook-processing logic is now contained within `Processor` instead of jumping around modules to see it you only have to look in one module

morriswchris commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+There is also a simple pipeline defined in `process` where it first checks if the webhook expired. If the webhook is expired, an error is returned and the pipeline is stopped so the webhook does not get sent out. If it isn't expired, then the webhook is sent to its endpoint in the next stage. Then the different responses returned by the endpoint are handled and saved to the database.
+
+The state needed for the webhook processing pipeline is defined at the top of the module, `WebhookState`, and it gets modified it as it moves through the different stages of the pipeline. For example, in the `send_webhook` stage after sending the webhook to its endpoint, the `response` is stored in the state so that the next stage, `handle_response`, can access it.
```
style: I believe you have an extra `it` here

> ... and it gets modified as it moves through the different stages of the pipeline

morriswchris commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+Lastly, the logic about not sending expired webhooks (`check_webhook_expired`) and saving different webhook endpoint responses to the database (`save_response_to_db`) was moved from the `Parser` and `Sender` modules into the `Processor` module. Now the `Parser` and `Sender` modules only contain logic related to their actual names.
+
+Now before we finish off by cleaning up the old modules, I had something to say about Elixir pipelines.
```
style: I think this should stay as present tense

> I ~~had~~ have something to say about ... 

You could also just drop this line altogether and just right into the next heading. Since this post reads like you would speak it is ok to keep though in my opinion

MichaelViveros commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+Some people inherit treasured jewelry, plots of land or even fortunes. Software engineers inherit spaghetti code written by people no longer working at the company that is need of more yak-shaving than the team of [grizzly Stanley Cup winners](https://www.urbandictionary.com/define.php?term=Playoff%20Beard) after their long road to victory. Which will undoubtedly be the Toronto Maple Leafs this y ...
+
+![Playoff Beard](https://s3951.pcdn.co/wp-content/uploads/2015/09/Brent-Burns-Sharks-2-575x431.jpg)
```
lol I personally like this one more too but think for people unfamiliar with the Leafs and even hockey in general the other picture is a better illustration of a grizzly beard desperately in need of a shave

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+It followed the *convenient* approach of spreading your business logic across lots of different parts of the application. This made it hard for new team members to get a big picture view of the service's application logic without travelling to the deep, dark depths of Mordor to figure out that the code that expires stale webhooks is embedded in the `Parser` module whose sole responsiblity should  ...
+
+Shortly before starting this project, my team was encouraged to take an excellent Elixir course that has unoficially become required training for all engineers at PagerDuty. This course is called ["Elixir for Programmers"](https://codestool.coding-gnome.com/courses/elixir-for-programmers) by Dave Thomas, aka The Coding Gnome (who I learned was also one of the authors of the timeless book, ["The P ...
```
Probably not unofficially required for _all_ back-end engineers, just those who write server-side code.

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+So, for this blog post, let's assume that the webhook processing service has the following business logic:
+1. Receive webhooks from an upstream service that generated them in response to an incident being triggered/acknowledged/resolved
+1. Parse the received webhooks from a JSON string
```
Some of your bullet points have a trailing period and some don't.

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+So, for this blog post, let's assume that the webhook processing service has the following business logic:
+1. Receive webhooks from an upstream service that generated them in response to an incident being triggered/acknowledged/resolved
+1. Parse the received webhooks from a JSON string
+1. Check if the webhooks are older than a certain threshold (aka expired). If they're expired, drop them.
```
"i.e. expired"

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+1. Parse the received webhooks from a JSON string
+1. Check if the webhooks are older than a certain threshold (aka expired). If they're expired, drop them.
+1. Send the webhooks to their endpoints via an HTTP POST request. These endpoints can be internal, customer-defined endpoints or external tools that PagerDuty integrates with like [Slack](https://www.pagerduty.com/docs/guides/slack-integration-guide/), [Jira](https://support.pagerduty.com/v1/docs/jira-cloud) and [ServiceNow](https://support.pagerduty.com/v1/docs/servicenow-integration-guide).
+1. Persist these webhook endpoint responses into a database (can be used to help debug customer issues).
```
Are the responses kept forever? Might make people nervous if so, but I'm not sure. If it's temporary, I'd call that out, otherwise not.

MichaelViveros replied:
Just above I explain that this is a simplification of our actual webhook processing system so I don't think I need to mention customer-considerations like that for this simplified version. And I think it might be confusing/contradicting if I say we're using a db but only store things temporarily.

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+In order to piece together that diagram above, you have to start in `Receiver`, go to `Parser`, go back to `Receiver`, go to `Sender` and finally finish off in `DB`.
+
+Along the way, you encounter some fundamental logic about how webhooks get processed which is hidden in modules that seemingly have nothing to do with that logic. As an example, the logic about not processing a webhook if it expired is contained in the `Parser` module whose sole responsibility should be decoding the webhook from JSON. Similarly, the logic about storing different types of webhook  ...
```
👏

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+Along the way, you encounter some fundamental logic about how webhooks get processed which is hidden in modules that seemingly have nothing to do with that logic. As an example, the logic about not processing a webhook if it expired is contained in the `Parser` module whose sole responsibility should be decoding the webhook from JSON. Similarly, the logic about storing different types of webhook  ...
+
+Overall, this webhook processing logic wouldn't be too hard to figure out for the simplified code above but for our application, which had a lot of additional complexity, it was quite daunting.
```
>Overall, this webhook processing logic wouldn't be too hard to figure out, given the simplified code above. But for our application -- which had a lot of additional complexity -- it was quite daunting.

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+Overall, this webhook processing logic wouldn't be too hard to figure out for the simplified code above but for our application, which had a lot of additional complexity, it was quite daunting.
+
+The code above also does not take advantage of the fact that webhook processing can be thought of as a pipeline which can be expressed beautifully in functional programming languages where you take in some initial state and keep modifying it through different stages to produce an output. Elixir implements a simple version of these pipelines via its [pipe operator (`|>`)](https://elixirschool.com/ ...
```
>... functional programming languages, where the code takes in some initial state and keeps modifying it ...

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+## The New Code
+
+So to centralize our webhook processing logic into one module and to express it as a pipeline, I moved all the bits of logic spread throughout the code into a new module called `Processor`. 
```
I'd drop the leading "so".

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+![New Business Logic](NewBusinessLogic.png)
+
+So with the new code, all of the webhook processing logic is now contained within `Processor` and instead of jumping around modules to see it you only have to look in one module.
```
Here too.

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+So with the new code, all of the webhook processing logic is now contained within `Processor` and instead of jumping around modules to see it you only have to look in one module.
+
+There is also a simple pipeline defined in `process` where it first checks whether the webhook expired. If the webhook is expired, an error is returned and the pipeline is stopped so the webhook does not get sent out. If it isn't expired, then the webhook is sent to its endpoint in the next stage. Then the different responses returned by the endpoint are handled and saved to the database.
```
👏

rwalbergpd commented on refactor_business_logic_more_functional.md:
````diff
@@ -1 +1,401 @@
+end
+```
+
+Once again, I wanted to lightly remind you that all of the code above is a simplification of our actual webhook processing pipeline. If I see any support requests (or even better, *tweets*) saying something like "Does PagerDuty seriously not even bother to retry sending failed webhooks???", I will simply smh. Then I will proceed to write a script that will page you every morning at 3 am the next  ...
````
This is funny and sounds like something I'd write but I'd get another opinion about this maybe being needlessly hostile. 😄

MichaelViveros replied:
Lol, being compared to your funny writing is a great compliment! I've gotten a few other engineers and some people from the Community team to read it over and they didn't have any reservations about this so should be all good.

rwalbergpd commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,401 @@
+
+![Refactoring](https://fiverr-res.cloudinary.com/images/t_main1,q_auto,f_auto/gigs/103760302/original/700668472c51ec832693a16397d638147e5452c3/refactor-your-code-in-a-day.jpg)
+
+Now I would be remiss not to mention that there are some benefits to embedding your business logic within different modules of your app. This can lead to less code and faster development and is one of the fundamental ideas behind frameworks like Ruby On Rails that prefer ["convention over configuration"](https://rubyonrails.org/doctrine/#convention-over-configuration). I personally disagree with  ...
```
Drop leading "now". You'd only say that in conversation.

MichaelViveros commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,427 @@
+
+![Parser Diff](ParserDiff.png)
+
+Once again, I wanted to lightly remind you that all of the code above is a simplification of our actual webhook processing pipeline. If I see any support requests (or even better, *tweets*) saying something like "Does PagerDuty seriously not even bother to retry sending failed webhooks???", I will simply smh. Then I will proceed to write a script that will page you every morning at 3 am the next  ...
```
Yeah, I see what you're saying. I've had some other engineers review this and most of them were fine with it. I could take out the last 2 sentences but I think that's the best part of the joke and I think this joke will make it that much more interesting/engaging.

But if you really feel strongly about it negatively impacting customer perceptions then I have no problem removing it.

MichaelViveros commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+(5 mins if you [agree](http://blog.firsthand.ca/2011/10/rails-is-not-your-application.html) that your application/business logic should be separate from everything else)
+
+(0 mins if you [disagree](https://twitter.com/dhh/status/218740829806792704) ;))
```
I meant it more as a subtle joke that if you disagree then you wouldn't even bother reading the blog. But I see how it can be confusing and I could link to these things at the end where I talk about some advantages of the contrasting opinion of not separating your business logic from everything else.

MichaelViveros commented on refactor_business_logic_more_functional.md:
```diff
@@ -1 +1,412 @@
+
+The decision to undertake this project as our next one was made for 2 main reasons:
+* To help the company [standardize on Elixir](https://www.pagerduty.com/blog/elixir-at-pagerduty/) due to an org-wide decision made last year
+* To give my team better visibility into how the service actually freakin' worked since reading 5+ year-old Scala written by engineers relatively new to the language resulted in files resembling this:
```
yeah, I thought that since this is going to be going up on the PD Eng blog (and not the main PD blog) I might have some more lee-way but I will definitely consider removing it
